### PR TITLE
Reenable PriorContext for Optimization

### PR DESCRIPTION
**Pull request by alyst**
This PR reenables the support for *PriorContext* in optimization, which was disabled by #2022.
While it is not directly supported by the current *optim_problem()* interface (only MLE/MAP are supported, although adding MaxPrior is possible), I think it is useful to support *PriorContext* for *Turing.OptimLogDensity()*.
In my code I use it for multi-objective Pareto optimization (prior + llh).

### CODE REVIEW

**Comment by devmotion** [comment]
> This PR reenables the support for _PriorContext_ in optimization, which was disabled by #2022.

AFAICT support for `PriorContext` has never existed and has never been tested. The check in the PR was only introduced to communicate this more clearly and provide a better error message.

**Comment by alyst** [comment]
> AFAICT support for PriorContext has never existed and has never been tested.

I see. It was working before though. Do you think it makes sense to support it for some of the user use cases?

**Comment by torfjelde** [comment]
IMO if we 're allowing both `DefaultContext` and `LikelihoodContext`, I feel like allowing `PriorContext` also seems reasonable (i.e. _all_ the leaf-contexts).

File: src/optimisation/Optimisation.jl [hunk 1]
```diff
@@ -83,6 +83,12 @@ function DynamicPPL.dot_tilde_assume(ctx::OptimizationContext, right, left, vns,
     return r, lp, vi
 end
 
+DynamicPPL.tilde_observe(::OptimizationContext{<:PriorContext}, right, left, vi) =
+    (0, vi)
+
+DynamicPPL.dot_tilde_observe(::OptimizationContext{<:PriorContext}, right, left, vi) =
+    (0, vi)
```
Comment (mhauru) [inline src/optimisation/Optimisation.jl:90]:
```suggestion
DynamicPPL.tilde_observe(ctx::OptimizationContext{<:PriorContext}, args...) =
    DynamicPPL.tilde_observe(ctx.context, args...)

DynamicPPL.dot_tilde_observe(ctx::OptimizationContext{<:PriorContext}, args...) =
    DynamicPPL.dot_tilde_observe(ctx.context, args...)
```
I think falling back on the `PriorContext` implementation is a bit neater. Result is the same.

**Review by mhauru — COMMENTED** [review]
I had one minor style change to suggest.

I also got some extra peace of mind by, in the test, checking that
```
            true_prior_logpdf = -Distributions.logpdf(Uniform(0, 2), a[1])
            @test Turing.OptimLogDensity(m1, prictx)(a) ≈ true_prior_logpdf
```
Could incorporate a more explicit check like that.

Both of these are optional, I'd be okay with the code as is. I'll leave approving for @torfjelde since there might be context (pun unintended) to this that I lack.

**Comment by alyst** [comment]
@mhauru I have incorporated your suggestions, thank you!

**Comment by torfjelde** [comment]
Would be happy to include this:) But we need to fix the merge conflicts :confused:

**Comment by mhauru** [comment]
Thanks @alyst!

**Comment by torfjelde** [comment]
Lovely work @alyst :)

**Pull request merged**